ddsrouter_event/include/ddsrouter_event/wait/impl/WaitHandler.ipp

```diff
@@ -0,0 +1,196 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file WaitHandler.ipp
+ */
+
+#include <ddsrouter_utils/Log.hpp>
+#include <ddsrouter_utils/Time.hpp>
+
+#ifndef _DDSROUTEREVENT_IMPL_WAITHANDLER_IPP_
+#define _DDSROUTEREVENT_IMPL_WAITHANDLER_IPP_
+
+namespace eprosima {
+namespace ddsrouter {
+namespace event {
+
+template <typename T>
+WaitHandler<T>::WaitHandler(
+        bool enabled /* = true */)
+    : enabled_(enabled)
+    , threads_waiting_(0)
+{
+}
+
+template <typename T>
+WaitHandler<T>::WaitHandler(
+        T init_value,
+        bool enabled /* = true */)
+    : enabled_(enabled)
+    , threads_waiting_(0)
+    , value_(init_value)
+{
+}
+
+template <typename T>
+WaitHandler<T>::~WaitHandler()
+{
+    blocking_disable();
+}
+
+template <typename T>
+void WaitHandler<T>::enable() noexcept
+{
+    // If disable, enable it. Otherwise do nothing
+    if(!enabled_.load())
+    {
+        // WARNING: enabled_ should be modified with mutex taken
+        std::lock_guard<std::mutex> lock(wait_condition_variable_mutex_);
+        logDebug(DDSROUTER_WAIT, "Enabling WaitHandler.");
+        enabled_.store(true);
+    }
+    else
+    {
+        logDebug(DDSROUTER_WAIT, "Enabling not disabled WaitHandler.");
+    }
+}
+
+template <typename T>
+void WaitHandler<T>::disable() noexcept
+{
+    // If enable, disable it. Otherwise do nothing
+    if(enabled_.load())
+    {
+        // WARNING: enabled_ should be modified with mutex taken
+        std::lock_guard<std::mutex> lock(wait_condition_variable_mutex_);
+        logDebug(DDSROUTER_WAIT, "Disabling WaitHandler.");
+        enabled_.store(false);
+
+        // Do not block for awaken
+        wait_condition_variable_.notify_all();
+    }
+    else
+    {
+        logDebug(DDSROUTER_WAIT, "Disabling not enabled WaitHandler.");
+    }
+}
+
+template <typename T>
+void WaitHandler<T>::blocking_disable() noexcept
+{
+    // Disable this object
+    disable();
+
+    // Wait till every thread has finished
+    while(threads_waiting_.load() > 0)
+    {
+        // Lock so this thread has no priority over closin threads
+        {
+            std::lock_guard<std::mutex> lock(wait_condition_variable_mutex_);
+        }
+
+        // Notify in case there are still threads waiting to be awaken
+        wait_condition_variable_.notify_all();
+    }
+}
+
+template <typename T>
+bool WaitHandler<T>::enabled() const noexcept
+{
+    return enabled_.load();
```

Comment:
Not protected, add _nts suffix

Reply:
Same with `get_value`

Reply:
I do not think they should be protected
And `_nts` methods should not be accesible from the user

Reply:
atomic::load is an atomic method, and it already returns a new copied value. This value is passed by reference at the return.
There is no need for a mutex here. (https://www.cplusplus.com/reference/atomic/atomic/load/)
Please, add an example where the mutex would be needed.

Reply:
You are right, but this limits the use of the template to trivially copyable types.